x/market/keeper/deal.go

```diff
@@ -125,3 +125,108 @@ func (k Keeper) SetDeal(ctx sdk.Context, deal types.Deal) {
+	err := k.isDataCertDuplicate(ctx, cert)
+	if err != nil {
+		return sdk.Coin{}, err
+	}
+
+	findDeal, err := k.GetDeal(ctx, cert.UnsignedCert.GetDealId())
+	if err != nil {
+		return sdk.Coin{}, err
+	}
+
+	if findDeal.GetStatus() == INACTIVE || findDeal.GetStatus() == COMPLETED {
+		return sdk.Coin{}, sdkerrors.Wrap(err, "deal is not activated")
+	}
+
+	dealAddress, err := types.AccDealAddressFromBech32(findDeal.GetDealAddress())
+	if err != nil {
+		return sdk.Coin{}, err
+	}
+
+	err = k.isTrustedValidator(cert, findDeal)
+	if err != nil {
+		return sdk.Coin{}, err
+	}
+
+	//TODO: Change the field(max_num_data) to price_per_data
+	totalAmount := findDeal.GetBudget().Amount.Uint64()
+	countOfData := findDeal.GetMaxNumData()
+
+	pricePerData := sdk.NewCoin("umed", sdk.NewIntFromUint64(totalAmount/countOfData))
+
+	dealBalance := k.bankKeeper.GetBalance(ctx, dealAddress, "umed")
+	if dealBalance.IsLT(pricePerData) {
+		return sdk.Coin{}, fmt.Errorf("deal's balance is smaller than reward")
+	}
+
+	coins := append(sdk.Coins{}, pricePerData)
+
+	err = k.bankKeeper.SendCoins(ctx, dealAddress, seller, coins)
+	if err != nil {
+		return sdk.Coin{}, sdkerrors.Wrapf(types.ErrNotEnoughBalance, "The deal's balance is not enough to make deal")
+	}
+
+	k.SetDataCertificate(ctx, findDeal.GetDealId(), cert)
+	SetCurNumData(&findDeal)
+	k.SetDeal(ctx, findDeal)
+	return pricePerData, nil
+
+}
+
+func (k Keeper) isDataCertDuplicate(ctx sdk.Context, cert types.DataValidationCertificate) error {
+	store := ctx.KVStore(k.storeKey)
+	dataCertKey := types.GetKeyPrefixCertificate(cert.UnsignedCert.GetDealId(), cert.UnsignedCert.GetDataHash())
+
+	if store.Has(dataCertKey) {
+		return fmt.Errorf("duplicated data")
+	}
+
+	return nil
+}
+
+func (k Keeper) isTrustedValidator(cert types.DataValidationCertificate, findDeal types.Deal) error {
+	validator := cert.UnsignedCert.GetDataValidatorAddress()
+	trustedValidators := findDeal.GetTrustedDataValidators()
+
+	for _, v := range trustedValidators {
+		if validator == v {
+			return nil
+		}
+	}
+	return fmt.Errorf("data validator is invalid address")
+}
+
+func (k Keeper) SetDataCertificate(ctx sdk.Context, dealId uint64, cert types.DataValidationCertificate) {
+	store := ctx.KVStore(k.storeKey)
+	dataHash := cert.UnsignedCert.GetDataHash()
+	dataCertificateKey := types.GetKeyPrefixCertificate(dealId, dataHash)
+	storedDataCertificate := k.cdc.MustMarshalBinaryLengthPrefixed(&cert)
+	store.Set(dataCertificateKey, storedDataCertificate)
+}
+
+func SetCurNumData(deal *types.Deal) {
+	curNumData := deal.GetCurNumData() + 1
+	deal.CurNumData = curNumData
+}
+
+func (k Keeper) Verify(ctx sdk.Context, validatorAddr sdk.AccAddress, cert types.DataValidationCertificate) (bool, error) {
+	validatorAcc := k.accountKeeper.GetAccount(ctx, validatorAddr)
+	if validatorAcc == nil {
+		return false, sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "invalid validator address")
+	}
```

Comment:
Yes. To be precise, if there isw no balance, the account itself does not exist on-chain, and no publicKey is created unless tx is performed.
This code will only response with an error if the account does not exist.
In my option, it would be good if the following code was added.
```go
if validatorAcc.GetPubKey() == nil {
  return false, sdkerrors.Wrapf(sdkerrors.ErrInvalidAddress, "the publicKey does not exist in the validator account")
}
```

Reply:
Yes, I agree with gyuguen's review. I think I need to add the validation code for the public key existed.